### PR TITLE
Leaf 3431 - New query parameter: Actionable

### PR DESCRIPTION
**Pull request by mgaoVA**
The new "actionable" query parameter identifies records that the current user has pending workflow actions. This inherits features from api/form/query and will be used as part of a performant Inbox.

Testing:
- Workflow access types: admin/non-admin/groups/designated person/designated group/requestor followup
- The workflow action box should have the same behavior as before
- Report Builder: setting "Current Status IS Actionable by me" should only show records that the current user has pending actions for
- Report Builder: setting "Current Status IS NOT Actionable by me" should only show resolved requests

### CODE REVIEW

File: LEAF_Request_Portal/FormWorkflow.php [hunk 1]
```diff
@@ -61,177 +63,206 @@ public function isActive(): bool
...
+
+            // dependencyID -1 is for a person designated by the requestor
+            if ($res[$i]['dependencyID'] == -1 && !$res[$i]['isActionable'])
+            {
+                $resEmpUID = $form->getIndicator($res[$i]['indicatorID_for_assigned_empUID'], 1, $res[$i]['recordID']);
 
-                // dependencyID 8 is for a special quadrad group
-                if ($res[$i]['dependencyID'] == 8 && !$res[$i]['hasAccess'])
+                // make sure the right person has access
+                if (!$res[$i]['isActionable'])
```
Comment (raphaelfretter) [inline LEAF_Request_Portal/FormWorkflow.php:146]: Already determined this is true in the parent if, why check again?

File: LEAF_Request_Portal/FormWorkflow.php [hunk 2]
```diff
@@ -61,177 +63,206 @@ public function isActive(): bool
...
-                $res[0]['filled'] = 0;
-            }
+        foreach ($res as $i => $record) {
+            // override access if user is in the admin group
+            $res[$i]['isActionable'] = $this->login->checkGroup(1); // initialize isActionable
 
-            for ($i = 0; $i < $numRes; $i++)
+            // check permissions
+            $res2 = $this->getDependency($res[$i]['dependencyID']);
+
```
Comment (raphaelfretter) [inline LEAF_Request_Portal/FormWorkflow.php:119]: Could just check if it is actionable at the top and continue if true.

**Review by raphaelfretter — CHANGES_REQUESTED** [review]

File: LEAF_Request_Portal/FormWorkflow.php (hunk 2 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/FormWorkflow.php:119]: Definitely more efficient

**Review by maxrichardjr — APPROVED** [review]

File: LEAF_Request_Portal/FormWorkflow.php [hunk 3]
```diff
@@ -61,177 +63,206 @@ public function isActive(): bool
...
+            $this->cache["dependencyID{$depID}"] = $this->db->prepared_query($strSQL, $vars);
         }
+        
+        return $this->cache["dependencyID{$depID}"];
+    }
 
-        $steps = array();
-        $vars = array(':recordID' => $this->recordID);
-        $strSQL = 'SELECT dependencyID, recordID, stepID, stepTitle, blockingStepID, workflowID, serviceID, filled, stepBgColor, stepFontColor, stepBorder, description, indicatorID_for_assigned_empUID, indicatorID_for_assigned_groupID, jsSrc, userID, requiresDigitalSignature FROM records_workflow_state
+    private function getServiceQuadrads(int $serviceID) {
```
Comment (Pelentan) [inline LEAF_Request_Portal/FormWorkflow.php]: Need return type specified.  Also, probably want some documentation on the function.

**Review by Pelentan — CHANGES_REQUESTED** [review]

**Review by jampaul3 — DISMISSED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by shaneodd — APPROVED** [review]

File: LEAF_Request_Portal/FormWorkflow.php (hunk 3 above)
Comment (mgaoVA) [reply LEAF_Request_Portal/FormWorkflow.php]: Updated

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by maxrichardjr — APPROVED** [review]

**Review by aerinkayne — APPROVED** [review]

**Review by raphaelfretter — DISMISSED** [review]

**Review by Pelentan — DISMISSED** [review]

**Review by AmandaSumner — DISMISSED** [review]

**Pull request merged**